File: core/src/net/natruid/jungle/utils/ai/BehaviorAction.java
package net.natruid.jungle.utils.ai;

import net.mostlyoriginal.api.event.common.EventSystem;
import net.natruid.jungle.components.BehaviorComponent;

public abstract class BehaviorAction extends BehaviorLeaf {
    protected EventSystem es;

    @Override
    public boolean run() {
        Float score = evaluate();
        if (score == null) return false;
        BehaviorComponent behaviorComponent = mBehavior.get(getSelf());
        if (score > behaviorComponent.score || behaviorComponent.execution == null) {
            behaviorComponent.score = score;
            behaviorComponent.execution = this;
        }
        return true;
    }

    public abstract Float evaluate();

    public abstract boolean execute();
}

File: core/src/net/natruid/jungle/utils/ai/BehaviorAction.kt
package net.natruid.jungle.utils.ai

import net.mostlyoriginal.api.event.common.EventSystem

abstract class BehaviorAction : BehaviorLeaf() {
    protected lateinit var es: EventSystem

    override fun run(): Boolean {
        val score = evaluate() ?: return false
        val behaviorComponent = mBehavior[self]
        if (score > behaviorComponent.score || behaviorComponent.execution == null) {
            behaviorComponent.score = score
            behaviorComponent.execution = this
        }
        return true
    }

    abstract fun evaluate(): Float?

    abstract fun execute(): Boolean
}
